### PR TITLE
Fix constraints for DEFAULT fields; fix constraints for temporary tables; improve performance of INSERT for small blocks.

### LINKED ISSUE #11273: Constraints don't apply to DEFAULT columns

### LINKED ISSUE #11274: Constraints are ignored for temporary tables

### LINKED ISSUE #11275: Suboptimal calculation of DEFAULT expressions if INSERT is done with small blocks via native protocol

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve performance for INSERT queries via INSERT SELECT or INSERT with clickhouse-client when small blocks are generated (typical case with parallel parsing). This fixes #11275. Fix the issue that CONSTRAINTs were not working for DEFAULT fields. This fixes #11273. Fix the issue that CONSTRAINTS were ignored for TEMPORARY tables. This fixes #11274.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
```
CREATE TABLE insert_small_block_performance
(
    `x` String, 
    `a` DEFAULT SHA256(x), 
    `b` DEFAULT SHA256(toString(a)), 
    `c` DEFAULT SHA256(toString(b)), 
    `d` DEFAULT SHA256(toString(c)), 
    `e` DEFAULT SHA256(toString(d)), 
    `f` DEFAULT SHA256(toString(e)), 
    `g` DEFAULT SHA256(toString(f)), 
    `h` DEFAULT SHA256(toString(g)), 
    `i` DEFAULT SHA256(toString(h)), 
    `j` DEFAULT SHA256(toString(i)), 
    `k` DEFAULT SHA256(toString(j)), 
    `l` DEFAULT SHA256(toString(k)), 
    `m` DEFAULT SHA256(toString(l)), 
    `n` DEFAULT SHA256(toString(m)), 
    `o` DEFAULT SHA256(toString(n)), 
    `p` DEFAULT SHA256(toString(o)), 
    `q` DEFAULT SHA256(toString(p)), 
    `r` DEFAULT SHA256(toString(q)), 
    `s` DEFAULT SHA256(toString(r)), 
    `t` DEFAULT SHA256(toString(s)), 
    `u` DEFAULT SHA256(toString(t)), 
    `v` DEFAULT SHA256(toString(u)), 
    `w` DEFAULT SHA256(toString(v))
)
ENGINE = Null
```

```
perl -e 'for (1..1000000) { print (("x" x 1000) . "\n") } ' | clickhouse-client --time --query "INSERT INTO insert_small_block_performance (x) FORMAT TSV"
```

master: 4.277 sec.
new: 2.267 sec.

**Comment by alexey-milovidov** [comment]
```
SET max_block_size = 1;
INSERT INTO insert_small_block_performance (x) SELECT toString(number) FROM numbers(1000000);
```

master: 484.56 rows/s.
new: 180.03 thousand rows/s.

File: src/Interpreters/InterpreterInsertQuery.cpp [hunk 1]
```diff
@@ -233,6 +233,21 @@ BlockIO InterpreterInsertQuery::execute()
...
+                    query.table_id, out, out->getHeader(), table->getConstraints(), context);
+
+            /// Actually we don't know structure of input blocks from query/table,
+            /// because some clients break insertion protocol (columns != header)
+            out = std::make_shared<AddingDefaultBlockOutputStream>(
+                out, query_sample_block, out->getHeader(), table->getColumns().getDefaults(), context);
+
+            /// It's important to squash blocks as early as possible (before other transforms),
+            ///  because other transforms may work inefficient if block size is small.
+
```
Comment (KochetovNicolai) [inline src/Interpreters/InterpreterInsertQuery.cpp:250]: So, there are two fixes here:
* defaults are calculating before constraints, so defaults are checked now
* squashing is performed first, and this improves performance
Looks fine.

File: src/Interpreters/InterpreterCreateQuery.cpp [hunk 2]
```diff
@@ -637,7 +637,7 @@ bool InterpreterCreateQuery::doCreateTable(ASTCreateQuery & create,
         if (create.if_not_exists && context.tryResolveStorageID({"", table_name}, Context::ResolveExternal))
             return false;
 
-        auto temporary_table = TemporaryTableHolder(context, properties.columns, query_ptr);
+        auto temporary_table = TemporaryTableHolder(context, properties.columns, properties.constraints, query_ptr);
```
Comment (KochetovNicolai) [inline src/Interpreters/InterpreterCreateQuery.cpp:640]: Constraints for temporary tables was just ignored. ok.

**Review by KochetovNicolai — APPROVED** [review]
Nice fix.

**Comment by alexey-milovidov** [comment]
Further info (ru):
```
1. После параллельного парсинга, клиент отправлял на сервер блоки меньшего размера.
2. Это не должно было повлиять, так как сервер склеивает блоки в более крупные при вставке.
3. Но при вставке, ещё производится вычисление DEFAULT выражений.
4. У клиента DEFAULT выражения очень забористые.
5. В коде порядок был неоптимальным - сначала вычисление DEFAULT, а потом склейка блоков в более крупные, а не наоборот.
6. Вычисление забористых выражений работает медленно, если блоки маленькие.
7. Попутно обнаружил, что также перепутан порядок проверки CONSTRAINTS, они тоже проверялись до DEFAULT, а не после.
8. В процессе написания теста пришлось добавить отсутствующую фичу - поддержка CONSTRAINTS для TEMPORARY TABLES (она просто отсутствовала, хотя это не баг).
9. На всякий случай, размер чанка для параллельного парсинга тоже увеличил в 10 раз, хотя это уже не так важно.
```

**Comment by alexey-milovidov** [comment]
> Stress test (thread) — Sanitizer assert (in stderr.log) 

Already fixed by @KochetovNicolai 

> Yandex synchronization check (only for Yandex employees) 
> Yandex third-party checks (only for Yandex employees) 

Broken by Yandex DevTools.

> Performance Pending — Started 

Very long, cannot wait.

**Pull request merged**